File: frontend/src/lib/spectrum/fitQuality.ts
// Fit-quality helpers for the EELS/EDS model-fit views (audit R4 / #2:
// ANALYSIS_PRESENTATION_PLAN.md item 2 "surface residuals + R² on
// existing fits"). Both /eels/fit and /eds/peakfit already return the
// measured spectrum ("spectrum") and the fitted curve ("model") over the
// SAME axis, so deriving the residual trace client-side is the thinnest
// correct path — serialising a third array over the wire would be a pure
// duplicate of `spectrum` and `model`, which are already there.
//
// `rSquared` is NOT how the R² shown in the UI is computed — that comes
// straight from the backend's `r_squared` field (calc/fit_quality.py),
// because the backend knows the ACTUAL fit window (e.g. EELS's
// `fit_range`), which can be a strict subset of the energy axis returned
// for plotting. Calling this on the full display axis would silently
// answer a different question ("how well does the fit extrapolate over
// the whole plot" rather than "how well did the fit explain what it was
// fit to"). It is exported and tested here anyway because it is the same
// pure formula the backend uses (`1 − SS_res/SS_tot`, plain and
// unweighted) — useful for any future client-only computation, and its
// own contract (in particular the degenerate SS_tot=0 case) is worth
// pinning with a test independent of a network round trip.

/** Point-wise "observed − model", the plotting convention (positive means
 *  the data sits above the fit). Throws on a length mismatch rather than
 *  silently truncating/padding — a shape mismatch means the caller passed
 *  two curves that were never fit against each other. */
export function residual(actual: number[], model: number[]): number[] {
  if (actual.length !== model.length) {
    throw new Error(
      `residual: actual (${actual.length}) and model (${model.length}) must be the same length`,
    );
  }
  return actual.map((a, i) => a - model[i]);
}

/** 1 − SS_res/SS_tot, mirroring calc/fit_quality.r_squared exactly
 *  (including its degenerate-SS_tot convention: a constant `actual`
 *  window has no variance to explain, so this returns 0 rather than
 *  NaN/±Infinity from a 0/0 division). */
export function rSquared(actual: number[], model: number[]): number {
  if (actual.length !== model.length) {
    throw new Error(
      `rSquared: actual (${actual.length}) and model (${model.length}) must be the same length`,
    );
  }
  const n = actual.length;
  if (n === 0) return 0;
  const mean = actual.reduce((sum, a) => sum + a, 0) / n;
  const ssTot = actual.reduce((sum, a) => sum + (a - mean) ** 2, 0);
  if (ssTot <= 0) return 0;
  const ssRes = actual.reduce((sum, a, i) => sum + (a - model[i]) ** 2, 0);
  return 1 - ssRes / ssTot;
}

File: src/fermiviewer/calc/fit_quality.py
"""Fit-quality scalar shared by the EELS/EDS model-fit routes (audit R4).

``calc/spectral_fit.py`` already computes ``reduced_chi2`` (a weighted
goodness-of-fit) and the raw ``residual``/``covariance``, but nothing
answers the plainer question a GUI readout wants: "how much of the
spectrum's own variance does the model explain?" That is the ordinary,
UNWEIGHTED coefficient of determination

    R^2 = 1 - SS_res / SS_tot
    SS_res = sum((actual - predicted)^2)
    SS_tot = sum((actual - mean(actual))^2)

— deliberately not weighted by the fit's 1/sigma^2 weights the way
``reduced_chi2`` is, so it stays a plain "fraction of variance explained"
number a non-statistician can read directly, matching the convention
already used by ``calc/ctf.py``'s ``CtfResult.r_squared``.

Callers MUST pass ``actual``/``predicted`` over the SAME window the fit
actually ran on (e.g. ``calc/eels_model.fit_edges``'s ``fit_range`` mask),
not a wider display range — R^2 answers "did the model explain what it
was fit to", and a window mismatch silently answers a different question.

Pure library: numpy only. No fastapi/pydantic/route imports (enforced by
tests/test_repo_integrity.py).
"""

from __future__ import annotations

import numpy as np

__all__ = ["r_squared"]


def r_squared(actual: np.ndarray, predicted: np.ndarray) -> float:
    """1 - SS_res/SS_tot over the given (already-windowed) arrays.

    Returns 0.0 when ``actual`` is degenerate (constant, so SS_tot == 0)
    instead of NaN/±inf — mirrors ``calc/ctf.py``'s ``r_squared``
    convention rather than leaving an undefined ratio for a flat/blank
    fit window.
    """
    a = np.asarray(actual, dtype=np.float64).ravel()
    p = np.asarray(predicted, dtype=np.float64).ravel()
    if a.shape != p.shape:
        raise ValueError("actual and predicted must be equal-length arrays")
    ss_tot = float(np.sum((a - a.mean()) ** 2))
    if ss_tot <= 0.0:
        return 0.0
    ss_res = float(np.sum((a - p) ** 2))
    return 1.0 - ss_res / ss_tot
